test/functional/eval/execute_spec.lua

```diff
@@ -33,14 +38,38 @@ describe('execute()', function()
     function! g:Bar()
       let a = ''
       redir => a
+      silent echon "bar1"
       call g:Foo()
+      silent echon "bar2"
       redir END
+      silent echon "bar3"
       return a
     endfunction
     ]])
-    eq('foo', funcs.execute('call g:Bar()'))
+    eq('top1bar1foobar2bar3', funcs.execute('echon "top1"|call g:Bar()'))
+  end)
 
-    eq('42', funcs.execute([[echon execute("echon execute('echon 42')")]]))
+  it('supports nested :redir to a register', function()
+    source([[
+    let @a = ''
+    function! g:Foo()
+      redir @a>>
+      silent echon "foo"
+      redir END
+      return @a
+    endfunction
+    function! g:Bar()
+      redir @a>>
+      silent echon "bar1"
+      call g:Foo()
+      silent echon "bar2"
+      redir END
+      silent echon "bar3"
+      return @a
+    endfunction
+    ]])
+    eq('top1bar1foobar2bar3', funcs.execute('echon "top1"|call g:Bar()'))
+    eq('bar1foobar2', eval('@a'))
```

Comment:
Ok. It's not something I care about, actually it was my misunderstanding of the existing functionality. 

Perhaps change this line so that it succeeds, with a comment mentioning that the behavior is expected.